api/include/opentelemetry/trace/global_provider.h

```diff
@@ -16,7 +16,8 @@ class DefaultTracerProvider final : public opentelemetry::trace::TracerProvider
 public:
   opentelemetry::trace::NoopTracer *const GetTracer(
       nostd::string_view library_name,
-      nostd::string_view library_version = "") override
+      nostd::string_view library_version  = "",
+      const TracerOptions &tracer_options = {}) override
```

Comment:
General C++ API question: constref means we can't meaningfully put a unique_ptr into TracerOptions to hand off ownership of an object. Should all options-accepting APIs take rvalues? e.g. `Thing(ThingOptions&& opts)`

@rnburn what do you think?

Reply:
Could follow the [javascript design](https://github.com/open-telemetry/opentelemetry-js/blob/d2e96ba3f90d53c51457cfcfd9a2362911323484/packages/opentelemetry-tracing/src/BasicTracerProvider.ts) where you can pass the default options to use in the constructor.

> constref means we can't meaningfully put a unique_ptr into TracerOptions to hand off ownership of an object

Yes, but if we want the options to be usable for constructing more than one tracer, I'm not sure we'd want them to use unique_ptrs

Reply:
> constructing more than one tracer

How common is this?

IMO most programs will have a single (and global) tracer. Requiring the exotic case to do slightly more work might be the right thing to do.

Reply:
I agree on defaulting to a single tracer.

Though, if you look at the default tracer-providers in other languages (e.g. [javascript](https://github.com/open-telemetry/opentelemetry-js/blob/d2e96ba3f90d53c51457cfcfd9a2362911323484/packages/opentelemetry-tracing/src/BasicTracerProvider.ts)), they construct a new tracer for each distinct (name,version) pair.

Right now, the PR is constructing a new tracer for each call to GetTracer ,which I think we should definitely change
https://github.com/lykkin/opentelemetry-cpp/blob/tracer-registry/sdk/src/opentelemetry/trace/tracer_provider.cc#L13-L14

Reply:
But then, the whole point of the [named tracers](https://github.com/open-telemetry/oteps/blob/master/text/0016-named-tracers.md) interface is to be able to construct more than one tracers.

My preference is to do this
1. Move TracerOptions out of the API (it's going to vary by vendor so it doesn't belong there).
2. Change the default TracerProvider provided by the SDK to behave like a singleton. Have it take TracerOptions in the constructor and make a single tracer that it returns for each call to GetTracer regardless of the name and version.

If users then want to use different tracers for different instrumented components in their application,  they can write their own TracerProvider that provides different tracers based on the name and version.

Reply:
> IMO most programs will have a single (and global) tracer

I think I misunderstood. Most *libraries* will have a single tracer per library. That's the point of adding the TraceProvider, right?

> Move TracerOptions out of the API

Agreed.

> Change the default TracerProvider provided by the SDK

Agreed.

And the default TracerProvider in the *API* should just hand out the same NoopTracer object?